File: backend/fastapi/app/services/rules.py

from datetime import date, timedelta
from fastapi import HTTPException
try:
    from dateutil import parser as dateparser
except ImportError:  # pragma: no cover - optional dependency
    dateparser = None
from ..models import CPRARequest, Timeline, TimelineItem

def to_date(s: str) -> date:
    """Convert a string to a ``date``.

    Supports ISO format (``YYYY-MM-DD``), US-style ``MM/DD/YYYY`` and falls back
    to ``dateutil.parser`` for more flexible parsing. Raises ``HTTPException`` if
    the format is unrecognized.
    """

    try:
        return date.fromisoformat(s)
    except ValueError:
        parts = s.split('/')
        if len(parts) == 3:
            try:
                month, day, year = (int(p) for p in parts)
                return date(year, month, day)
            except ValueError:
                pass
        if dateparser is not None:
            try:
                return dateparser.parse(s).date()
            except (ValueError, OverflowError):
                pass
        raise HTTPException(status_code=400, detail=f"Unrecognized date format: {s}")

def roll_forward(d: date) -> date:
    if d.weekday() == 5: return d + timedelta(days=2) # Sat->Mon
    if d.weekday() == 6: return d + timedelta(days=1) # Sun->Mon
    return d

def compute_timeline(req: CPRARequest, adjust_for_holidays: bool=True) -> Timeline:
    received = to_date(req.receivedDate)
    determination = received + timedelta(days=10)
    extension_due = None
    if req.extension.apply:
        extension_due = determination + timedelta(days=14)
    if adjust_for_holidays:
        determination = roll_forward(determination)
        if extension_due: extension_due = roll_forward(extension_due)
    prod_base = extension_due or determination
    milestones = [
        TimelineItem(label="Search kickoff", due=str(roll_forward(received + timedelta(days=1)))),
        TimelineItem(label="Privilege review", due=str(roll_forward(prod_base - timedelta(days=3)))),
        TimelineItem(label="Draft production", due=str(roll_forward(prod_base - timedelta(days=1)))),
    ]
    return Timeline(determinationDue=str(determination), extensionDue=(str(extension_due) if extension_due else None), milestones=milestones)

File: backend/express/src/services/rules.ts
import type { CPRARequest, Timeline, TimelineItem } from './types.js';

function toDate(s: string): Date {
  const iso = new Date(s);
  if (!isNaN(iso.getTime())) return new Date(iso.getFullYear(), iso.getMonth(), iso.getDate());
  const parts = s.split('/');
  if (parts.length === 3) {
    const [month, day, year] = parts.map(p => parseInt(p, 10));
    const d = new Date(year, month - 1, day);
    if (!isNaN(d.getTime())) return d;
  }
  const parsed = Date.parse(s);
  if (!isNaN(parsed)) {
    const d = new Date(parsed);
    return new Date(d.getFullYear(), d.getMonth(), d.getDate());
  }
  throw new Error(`Unrecognized date format: ${s}`);
}

function rollForward(d: Date): Date {
  const day = d.getDay();
  if (day === 6) return new Date(d.getTime() + 2 * 86400000);
  if (day === 0) return new Date(d.getTime() + 1 * 86400000);
  return d;
}

function fmt(d: Date): string {
  return d.toISOString().slice(0, 10);
}

export function computeTimeline(req: CPRARequest, adjustForHolidays = true): Timeline {
  if (!req || !req.requester || !req.receivedDate || !req.description) {
    throw new Error('Invalid CPRARequest');
  }
  const received = toDate(req.receivedDate);
  let determination = new Date(received.getTime() + 10 * 86400000);
  let extensionDue: Date | undefined;
  if (req.extension && req.extension.apply) {
    extensionDue = new Date(determination.getTime() + 14 * 86400000);
  }
  if (adjustForHolidays) {
    determination = rollForward(determination);
    if (extensionDue) extensionDue = rollForward(extensionDue);
  }
  const prodBase = extensionDue ? extensionDue : determination;
  const milestones: TimelineItem[] = [
    { label: 'Search kickoff', due: fmt(rollForward(new Date(received.getTime() + 1 * 86400000))) },
    { label: 'Privilege review', due: fmt(rollForward(new Date(prodBase.getTime() - 3 * 86400000))) },
    { label: 'Draft production', due: fmt(rollForward(new Date(prodBase.getTime() - 1 * 86400000))) },
  ];
  return {
    determinationDue: fmt(determination),
    extensionDue: extensionDue ? fmt(extensionDue) : undefined,
    milestones,
  };
}
